### ISSUE TITLE
PHPUnit Tests bundled with Magento fatal errors

### ISSUE THREAD

**Issue by lbajsarowicz** [magento/magento2#27638]
### Preconditions (*)
1. Vanilla Magento 2.3.4

### Steps to reproduce (*)
1. `vendor/bin/phpunit -c dev/tests/phpunit.xml.dist`

### Expected result (*)
1. Tests executed till the end

### Actual result (*)
1. Fatal errors:
```
Fatal error: Class Mock_CartExtensionInterface_0ba80a0b contains 2 abstract methods and must therefore be declared abstract or implement the remaining methods (Magento\Quote\Api\Data\CartExtensionInterface::getAmazonOrderReferenceId, Magento\Quote\Api\Data\CartExtensionInterface::setAmazonOrderReferenceId) in /var/www/html/vendor/phpunit/phpunit-mock-objects/src/Generator.php(264) : eval()'d code on line 1
```

**Comment by lbajsarowicz** [comment]
As I was asked to update the issue:
![image](https://user-images.githubusercontent.com/1639941/79020988-ac969e00-7b7a-11ea-8f39-698ca897c46c.png)

```
PHP Fatal error:  Class Magento\Catalog\Api\Data\ProductExtension contains 18 abstract methods and must therefore be declared abstract or implement the remaining methods (Magento\Catalog\Api\Data\ProductExtensionInterface::getWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::setWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::getCategoryLinks, ...) in /var/www/html/dev/tests/unit/tmp/generated/code/Magento/Catalog/Api/Data/ProductExtension.php on line 4

Fatal error: Class Magento\Catalog\Api\Data\ProductExtension contains 18 abstract methods and must therefore be declared abstract or implement the remaining methods (Magento\Catalog\Api\Data\ProductExtensionInterface::getWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::setWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::getCategoryLinks, ...) in /var/www/html/dev/tests/unit/tmp/generated/code/Magento/Catalog/Api/Data/ProductExtension.php on line 4
```

PHP:
```
www-data@rescue-php-fpm:08:28 PM:/var/www/html$ php -v
PHP 7.3.15 (cli) (built: Mar  9 2020 03:13:18) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.15, Copyright (c) 1998-2018 Zend Technologies
    with the ionCube PHP Loader + ionCube24 v10.3.9, Copyright (c) 2002-2019, by ionCube Ltd.
    with Zend OPcache v7.3.15, Copyright (c) 1999-2018, by Zend Technologies

```

**Comment by lbajsarowicz** [comment]
Another one <3
```
Test 'Magento\Catalog\Test\Unit\Model\Product\CopierTest::testCopy' started
PHP Fatal error:  Class Magento\Catalog\Api\Data\ProductExtension contains 18 abstract methods and must therefore be declared abstract or implement the remaining methods (Magento\Catalog\Api\Data\ProductExtensionInterface::getWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::setWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::getCategoryLinks, ...) in /var/www/html/dev/tests/unit/tmp/generated/code/Magento/Catalog/Api/Data/ProductExtension.php on line 4

Fatal error: Class Magento\Catalog\Api\Data\ProductExtension contains 18 abstract methods and must therefore be declared abstract or implement the remaining methods (Magento\Catalog\Api\Data\ProductExtensionInterface::getWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::setWebsiteIds, Magento\Catalog\Api\Data\ProductExtensionInterface::getCategoryLinks, ...) in /var/www/html/dev/tests/unit/tmp/generated/code/Magento/Catalog/Api/Data/ProductExtension.php on line 4
```

**Comment by lbajsarowicz** [comment]
Ooo! Who could expect.
Another one!
```

There was 1 error:

1) Magento\Framework\Stdlib\Test\Unit\DateTime\TimezoneTest::testDateIncludeTime with data set "Parse greek d/m/y date with time" ('30/10/2021, 12:01 π.μ.', 'el_GR', true, 1635570060)
Exception: DateTime::__construct(): Failed to parse time string (30/10/2021, 12:01 π.μ.) at position 0 (3): Unexpected character

/var/www/html/lib/internal/Magento/Framework/Stdlib/DateTime/Timezone.php:186
/var/www/html/lib/internal/Magento/Framework/Stdlib/Test/Unit/DateTime/TimezoneTest.php:95
```

**Comment by ihor-sviziev** [comment]
I see that this issue was already fixed by #27701, so I'm closing this issue.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz. Thank you for your report. 
The issue has been fixed in magento/magento2#27701 by @lbajsarowicz in 2.4-develop branch
Related commit(s):
 - [f34f8c3e3a775a0f67ab4eb5b1edea5dc83ee5c7](https://github.com/magento/magento2/commit/f34f8c3e3a775a0f67ab4eb5b1edea5dc83ee5c7)
 - [f616fd85974a003d1aa8bda9ef3009ab3ec6030b](https://github.com/magento/magento2/commit/f616fd85974a003d1aa8bda9ef3009ab3ec6030b)
 - [dd39f7ef6991820a861b3a22276c8c82ccc816e1](https://github.com/magento/magento2/commit/dd39f7ef6991820a861b3a22276c8c82ccc816e1)
 - [8e30aa9cf2ceb0d6e56eebaf47c0cf629877f0f7](https://github.com/magento/magento2/commit/8e30aa9cf2ceb0d6e56eebaf47c0cf629877f0f7)

The fix will be available with the upcoming 2.4.0 release.

**Comment by jacekl** [comment]
Good job @lbajsarowicz !

I have the same issue upgrading from 2.3.3. to 2.3.5... One thing to mention:
* it happens after you run `setup:di:compile` command before running unit tests,
* if you remove `generated/code` directory contents and run unit tests without `setup:di:compile` - tests are execuring correctly.

At least for me.
